File: mobile/lib/features/areas/domain/entities/area.dart
import 'package:equatable/equatable.dart';
import 'area_component_binding.dart';
import 'area_status.dart';

class Area extends Equatable {
  final String id;
  final String name;
  final String? description;
  final AreaStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final AreaComponentBinding action;
  final List<AreaComponentBinding> reactions;

  const Area({
    required this.id,
    required this.name,
    required this.description,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.action,
    required this.reactions,
  });

  bool get isEnabled => status == AreaStatus.enabled;

  Area copyWith({
    String? id,
    String? name,
    String? description,
    AreaStatus? status,
    DateTime? createdAt,
    DateTime? updatedAt,
    AreaComponentBinding? action,
    List<AreaComponentBinding>? reactions,
  }) {
    return Area(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      action: action ?? this.action,
      reactions: reactions ?? this.reactions,
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    description,
    status,
    createdAt,
    updatedAt,
    action,
    reactions,
  ];
}

File: server/internal/domain/area/area.go
package area

import (
	"strings"
	"time"

	"github.com/google/uuid"
)

// Status mirrors the area_status enum from the persistence layer
type Status string

const (
	// StatusEnabled indicates the automation is active and should run
	StatusEnabled Status = "enabled"
	// StatusDisabled prevents the automation from executing while retaining configuration
	StatusDisabled Status = "disabled"
	// StatusArchived marks the automation as read-only and hidden from default listings
	StatusArchived Status = "archived"
)

// Area represents an automation composed of an action and one or more reactions
type Area struct {
	ID          uuid.UUID
	UserID      uuid.UUID
	Name        string
	Description *string
	Status      Status
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// WithDescription returns a copy of the area with the provided description applied
func (a Area) WithDescription(description string) Area {
	if strings.TrimSpace(description) == "" {
		a.Description = nil
		return a
	}
	desc := description
	a.Description = &desc
	return a
}

// OwnedBy reports whether the area belongs to the provided user
func (a Area) OwnedBy(userID uuid.UUID) bool {
	return a.UserID == userID
}

// WithStatus returns a copy of the area carrying the provided status
func (a Area) WithStatus(status Status) Area {
	a.Status = status
	return a
}

// WithTimestamps sets the created and updated timestamps if provided
func (a Area) WithTimestamps(created time.Time, updated time.Time) Area {
	a.CreatedAt = created
	a.UpdatedAt = updated
	return a
}
